### DISCUSSION TITLE
No audio from HFP

### DISCUSSION THREAD

**Discussion by david-bouthillier** [Arkq/bluez-alsa#690] [Q&A] [answered]
My goal is to make a phone call using an embedded device through my Bluetooth connected phone. Playing audio off the phone with the A2DP profile works, but when I trying to make a call, there is no audio.

I tried to play the phone call audio with this script:

`alsaloop -C bluealsa:DEV=<my_bluetooth_addr>,PROFILE=sco -P hw:0 -r 8000 -c 2 -f s16_le -n -t 40000 -v`

Here is some information about my system. I'm working on a Toradex iMX8MP Verdin board.

```
root@verdin-imx8mp-15006329:~# tdx-info
Software summary
Bootloader: U-Boot
Kernel version: 5.15.129-6.4.0-devel+git.67c3153d20ff #1 SMP PREEMPT Wed Sep 27 12:30:36 UTC 2023
Kernel command line: root=PARTUUID=c43924ef-02 ro rootwait console=tty1 console=ttymxc2,115200 consoleblank=0 earlycon
Distro name: NAME=“TDX Wayland with XWayland”
Distro version: VERSION_ID=6.4.0-devel-20231220205908-build.0
Hostname: verdin-imx8mp-15006329

Hardware info
HW model: Toradex Verdin iMX8M Plus WB on Verdin Development Board
Toradex version: 0058 V1.1A
Serial number: 15006329
Processor arch: aarch64
```

```
root@verdin-imx8mp-15006329:~# bluealsa --version
v4.0.0
```

```
root@verdin-imx8mp-15006329:~# bluealsa-aplay -L
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=a2dp
    David iPhone, phone, capture
    A2DP (SBC): S16_LE 2 channels 44100 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, capture
    SCO (CVSD): S16_LE 1 channel 8000 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, playback
    SCO (CVSD): S16_LE 1 channel 8000 Hz
```

```
root@verdin-imx8mp-15006329:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: imx8mpnau8822 [imx8mp-nau8822], device 0: 30c10000.sai-nau8822-hifi nau8822-hifi-0 [30c10000.sai-nau8822-hifi nau8822-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: audiohdmi [audio-hdmi], device 0: i.MX HDMI i2s-hifi-0 [i.MX HDMI i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

```
root@verdin-imx8mp-15006329:~# arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: imx8mpnau8822 [imx8mp-nau8822], device 0: 30c10000.sai-nau8822-hifi nau8822-hifi-0 [30c10000.sai-nau8822-hifi nau8822-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

I read the wiki section about SCO routing. When I ran the example I DID NOT hear any audio AND the TX SCO count remained at 0. The last paragraph of that section mentions that audio is not passing from my application to BlueALSA. Can someone help me diagnose this? I did verify that I could play the audio with a simple `aplay` command, so the sound card is working.

https://github.com/arkq/bluez-alsa/wiki/Using-BlueALSA-with-HFP-and-HSP-Devices#the-sco-routing-issue

Any help is greatly appreciated!

**Comment by borine** [comment]
>  I DID NOT hear any audio AND the TX SCO count remained at 0

I'm not sure what you mean by that. Are you saying that there was a call in progress and the phone was sending the audio stream of the call to/from the bluealsa device; and while that was happening you tried something like:
```
aplay -D bluealsa:PROFILE=sco /usr/share/sounds/alsa/Front_Centre.wav
```
but the sco TX count remained at zero ?

**Reply by david-bouthillier** [reply]
No, there was no call in progress. I was connect to my phone via bluetooth. I ran the aplay test on my embedded device, like the wiki document mentions. I did not hear any sound, and the counters did not increase.

Have I misunderstood something?

**Reply by arkq** [reply]
If you want to stream audio from a device to iPhone, you have to have ongoing call or start some recording application which can use Bluetooth headset's microphone (in that case the headset is your device with BlueALSA). I'm not 100%, but I think that iOS built-in recording app uses Bluetooth for recording (that's not always the case with Android, though).

**Reply by borine** [reply]
When you *do* have an ongoing call, then the SCO RX bytes counter should increase because the adapter should be sending the incoming SCO audio over the HCI which will therefore appear on the HCI RX bytes counter. If the sco RX bytes is at zero, that means the adapter is not using the HCI for SCO audio. I note that the Verdin board uses an adapter based on a Marvell module which has PCM and I2S interfaces in addition to the USB and UART ones. So perhaps SCO audio is being routed over one of those, in which case bluealsa cannot access it.

**Reply by david-bouthillier** [reply]
@borine the folks at Toradex have pointed me in this direction as well. Looks like it's an issue getting SCO packets through the HCI interface. They provided a solution that allows the call audio through, but the headset mic doesn't make its way back to the phone.

**Reply by borine** [reply]
Could you try using msbc codec instead of cvsd? That may then use the HCI if it requires the host to decode the mabc frames.

**Reply by david-bouthillier** [reply]
@borine We recompiled blueASLA with mSBC support, but how do I select that codec for SCO profile?

**Reply by arkq** [reply]
In HFP the audio gateway selects profile, which in your case is iPhone. Before establishing SCO connection it should select mSBC codec (if available). If sampling frequency is 16000 Hz it means that mSBC was selected.

**Reply by david-bouthillier** [reply]
@arkq It does select mSBC after I initiate a call on my phone. I still cannot hear any audio though.

```
root@verdin-imx8mp-15007248:~# bluealsa-aplay -L
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=a2dp
    David iPhone, phone, capture
    A2DP (SBC): S16_LE 2 channels 44100 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, capture
    SCO (<null>): S16_LE 1 channel 0 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, playback
    SCO (<null>): S16_LE 1 channel 0 Hz
root@verdin-imx8mp-15007248:~# bluealsa-aplay -L
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=a2dp
    David iPhone, phone, capture
    A2DP (SBC): S16_LE 2 channels 44100 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, capture
    SCO (mSBC): S16_LE 1 channel 16000 Hz
bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco
    David iPhone, phone, playback
    SCO (mSBC): S16_LE 1 channel 16000 Hz
```

**Reply by borine** [reply]
>  I still cannot hear any audio though

So some more questions to help us understand your setup:

* What is this "solution" provided by Toradex ?
* Is the incoming audio from the phone playing through `alsaloop` from bluealsa?
* What command line are you using to get the sound from the mic to bluealsa? (and does that command produce any error or debug messages?

It would also be good to see the output of `hciconfig -a` just to see some more detail about your adapter.

**Reply by david-bouthillier** [reply]
@borine Toradex provided me the following command to run to adjust SCO routing. This fixed the audio for the phone call, but the microphone audio is not passed back from the embedded device to the phone.

`hcitool -i hci0 cmd 0x3F 0x001D 0x00`

Yes, the incoming audio is playing through alsaloop. Here is the full script I run, which includs the command for the mic. There are no errors that I have seen.

```
alsaloop -g /dev/stdin <<EOF
-C bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco -P plughw:0 -r 8000 -c 1 -f s16_le -n -t 40000 -T 1
-C plughw:0 -P bluealsa:SRV=org.bluealsa,DEV=A0:56:F3:97:94:4C,PROFILE=sco -r 8000 -c 1 -f s16_le -n -t 40000 -T 2
EOF
```

Here is the output of hciconfig -a

```
root@verdin-imx8mp-15007248:~# hciconfig -a
hci0:	Type: Primary  Bus: UART
	BD Address: E8:FB:1C:CD:E1:64  ACL MTU: 1021:5  SCO MTU: 60:12
	UP RUNNING 
	RX bytes:79973 acl:95 sco:0 events:1829 errors:0
	TX bytes:3443 acl:83 sco:0 commands:117 errors:0
	Features: 0xff 0xfe 0x8f 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH HOLD SNIFF 
	Link mode: PERIPHERAL ACCEPT 
	Name: 'verdin-imx8mp-15007248'
	Class: 0x2c0000
	Service Classes: Rendering, Capturing, Audio
	Device Class: Miscellaneous, 
	HCI Version: 5.2 (0xb)  Revision: 0x8300
	LMP Version: 5.2 (0xb)  Subversion: 0x1454
	Manufacturer: Marvell Technology Group Ltd. (72)
```

**Reply by borine** [reply]
I guess the hciconfig output was taken before running the tests. Are both sco RX and sco TX bytes greater than zero after the test?

If so, and the alsaloop command produced no error messages, then it looks like the mic audio is correctly flowing through alsaloop and bluealsa to the bluetooth adapter.

If sco TX bytes remains at zero, then bluealsa is not sending the audio from alsaloop to the HCI. In that case we would need debug logs from bluealsa to see what is going on. If the TX bytes grows steadily while the test is running, then the error must be in the adapter or the phone.

**Reply by david-bouthillier** [reply]
@borine here is the output after making a call and then running hciconfig -a

```
root@verdin-imx8mp-15007248:~# hciconfig -a
hci0:	Type: Primary  Bus: UART
	BD Address: E8:FB:1C:CD:E1:64  ACL MTU: 1021:5  SCO MTU: 60:12
	UP RUNNING 
	RX bytes:182639 acl:104 sco:2134 events:1053 errors:0
	TX bytes:131732 acl:85 sco:2004 commands:114 errors:0
	Features: 0xff 0xfe 0x8f 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH HOLD SNIFF 
	Link mode: PERIPHERAL ACCEPT 
	Name: 'verdin-imx8mp-15007248'
	Class: 0x2c0000
	Service Classes: Rendering, Capturing, Audio
	Device Class: Miscellaneous, 
	HCI Version: 5.2 (0xb)  Revision: 0x8300
	LMP Version: 5.2 (0xb)  Subversion: 0x1454
	Manufacturer: Marvell Technology Group Ltd. (72)
```

I could hear audio, but my mic audio wasn't picked up.
There were no errors from alsaloop or in the system log.

**Reply by borine** [reply, marked as answer]
Well, it is now clear that bluealsa is sending audio to the HCI. So some possibilities I can think of:
* the mic is muted and alsaloop is capturing only silence - check the controls for the card with alsamixer.
* the bluealsa SCO playback pcm is muted and sending only silence - check the controls for bluealsa with alsamixer
* the adapter is not processing SCO data received from the HCI - check with Verdin support
* the phone is not connecting the incoming bluetooth audio to an active call (unlikely !!)

**Reply by david-bouthillier** [reply]
@borine Thanks so much for your assistance. We ran a test on our mic and it turns out the unit wasn't working!

So if anyone else ends up here in the future, the solution was the following:
- Run hcitool to properly configure SCO packets to be routed over HCI (this command appears to be chip specific)
- Make sure your mic is working!
